### ISSUE TITLE
Numerically stable n−1 evaluation and removal of dead _utils helpers (A4, R9)

### ISSUE THREAD

**Issue by chrisfinlay** [chrisfinlay/jax-nufft#12]
Part of https://github.com/chrisfinlay/jax-nufft/issues/2 · Tracking: #35 · Depends on: none

## Context
`planning.make_plan` computes `sqrt(1 − l² − m²) − 1`, which cancels catastrophically near the phase centre
(exactly 0 for the innermost pixels in float32). The cancellation-free form is the standard rewrite
`√(1−x) − 1 = −x / (√(1−x) + 1)` ([Higham2002] Ch. 1). `src/jax_nufft/_utils.py::n_minus_1_grid`,
`uvw_to_finufft_coords` and `uvw_meters_to_lambda` are unused anywhere in `src/` or `tests/`.

## Implementation plan
1. `planning.make_plan`: compute `inside_val = -eps_lm / (sqrt(1 - eps_lm) + 1)` for `eps_lm <= 1`, keep the analytic
   extension outside the disc. Compute in float64 regardless of the plan dtype and cast at the end.
2. Delete the three unused helpers from `_utils.py` (keep `SPEED_OF_LIGHT`); grep confirms no importers.
3. `tests/test_planning.py`: add `test_nm1_stable_near_centre` comparing `plan.n_minus_1` against the stable form
   evaluated with `numpy.longdouble` at the 3×3 innermost pixels, relative error < 1e-12 for a float64 plan and
   < 1e-6 for a float32 plan (once the dtype PR has landed; otherwise float64 only).

## Definition of done
- [ ] New test passes; existing planning tests unchanged.
- [ ] `_utils.py` contains only what is imported.

## Working context (self-contained)

- **Environment.** `pixi run -e test <cmd>` from the repo root (Python 3.12; jax, jax-finufft, numpy; ducc0 as the test
  oracle). Read `AGENTS.md` §1–§6 first: sign convention, plan-then-call pattern and the plan-field checklist (§4),
  strategies (§5), test/benchmark commands (§6). `pixi run -e test pytest -q --runslow`, `pixi run -e dev lint` and
  `pixi run -e dev typecheck` must stay green.
- **Fixtures.** `tests/conftest.py` defines telescopes `EDA2` (64², 120° FoV, 400 rows), `MWA_COMPACT` (128²),
  `MWA_EXTENDED` (256²), `MEERKAT` (256²), `GH200_LARGE` (2048², 50k rows) and `synthetic_uvw(tel, zenith_angle_deg, seed)`.
  "`<name>` zenith / off30" means `uvw = synthetic_uvw(tel, 0.0 or 30.0, seed=0)`, `freq = np.array([tel.freq_hz])`,
  `image_shape = (tel.n_pix, tel.n_pix)`, `pixsize = tel.pixsize`, image `np.random.default_rng(7).standard_normal(...)`
  and complex vis from the same generator. The "seven review fixtures" are EDA2 zenith, MWA_compact zenith and off30,
  MWA_extended zenith and off30, MeerKAT zenith and off30. The "five timing fixtures" are MWA_compact off30,
  MWA_extended off30, MeerKAT off30, EDA2 zenith, MWA_extended zenith.
- **Accuracy metric.** Relative L2 error `‖a − b‖₂ / ‖b‖₂` against the exact DFT in the repo's sign convention
  (`tests/test_against_dft.py::_reference_forward`, `tests/test_adjoint.py::_reference_adjoint`; vectorise over row
  chunks to make 256² fixtures take seconds). "x× eps" is that ratio divided by the requested epsilon. "The accuracy
  sweep" is the seven review fixtures × eps ∈ {1e-3, 1e-4, 1e-5, 1e-6, 1e-7, 1e-8, 1e-10, 1e-12}, forward and adjoint,
  `dense_scan` (the other strategies agree with it to 1e-14). The width-rule sub-issue adds
  `tests/test_accuracy_sweep.py` (gated by `--runsweep`) that produces this table; later issues re-run it.
- **ducc0 as oracle.** `ducc0.wgridder.dirty2vis(uvw=, freq=, dirty=, pixsize_x=, pixsize_y=, epsilon=, do_wgridding=True,
  divide_by_n=False, nthreads=1)` and `ducc0.wgridder.vis2dirty(..., npix_x=, npix_y=, divide_by_n=True)` match
  jax-nufft's default conventions (see `tests/test_against_ducc.py`). With `verbosity=1` ducc0 prints its plane count to
  C-level stdout as `(nu x nv x nplanes), supp=W`; capture it by redirecting file descriptor 1 (`os.dup2` to a temp file,
  not `contextlib.redirect_stdout`) and parse `r"\((\d+)x(\d+)x(\d+)\), supp=(\d+)"`.
- **Timing protocol.** Same machine before and after; `nthreads` stated; plan built and one warm-up call done outside the
  timer; median of ≥ 7 calls, each ending with `.block_until_ready()`. Timing gates are relative to the pre-PR number
  measured this way; absolute times quoted above are from the review machine (Apple M-series, 10 cores) for orientation only.
- **Memory protocol.** `jax.jit(fn).lower(*args).compile().memory_analysis().temp_size_in_bytes` (works on the CPU backend);
  compare with the same quantity for the plain forward/adjoint call.
- **GPU gates.** Need the CUDA jax-finufft build (pixi env `gpu-dev`, Linux). Without a GPU: satisfy the CPU gates, state in
  the PR that the GPU gate is pending for the maintainer, and use `docs/benchmarks/v0.1.2-baseline-gpu.json` (GH200
  per-cell median seconds, `n_w`, padding overhead) as the reference numbers.
- **PR conventions.** One PR per issue, branched from `dev`, title `<short> (closes #N)`, the definition-of-done checklist
  copied into the description with before/after numbers, and the statement that ducc0 source was not consulted.

Review finding IDs (private review document): A4, R9.

---
**Sources to implement from.** ducc0 is GPL-2.0-or-later; it is used in this project only as a black-box test
oracle through its public Python API. Do not consult or transliterate its source for this work; implement from the
mathematics stated above and these references (FINUFFT and jax-finufft are Apache-2.0 and may be read freely):

- [Arras+2021] P. Arras, M. Reinecke, R. Westermann, T. A. Enßlin, *Efficient wide-field radio interferometry response*, A&A 646, A58 (2021), arXiv:2010.10122 — §2 w-stacking, eq. 12 plane spacing, eq. 19 Hermitian symmetry, eqs. 20–21 plane placement, eq. 29 modified ES kernel, §3.2 + Appendix A optimal kernel parameters, §4.1 cost model.
- [Barnett+2019] A. H. Barnett, J. Magland, L. af Klinteberg, *A parallel nonuniform fast Fourier transform library based on an "exponential of semicircle" kernel*, SIAM J. Sci. Comput. 41(5), C479 (2019), arXiv:1808.06736 — eq. 8 ES kernel, eq. 10 β = 2.30·W and the width rule, Theorem 7 error decay, eq. 18 Gauss–Legendre kernel transform.
- [Ye+2022] H. Ye, S. F. Gull, S. M. Tan, B. Nikolic, *Optimal gridding and degridding in radio interferometry imaging*, MNRAS 510, 4110 (2022), arXiv:2110.03914 — kernel optimisation methodology.
- [TMS2017] A. R. Thompson, J. M. Moran, G. W. Swenson, *Interferometry and Synthesis in Radio Astronomy*, 3rd ed., Springer (2017), Ch. 3 — conjugate symmetry of the visibility function and the shift theorem.
- [Higham2002] N. J. Higham, *Accuracy and Stability of Numerical Algorithms*, 2nd ed., SIAM (2002), Ch. 1 — cancellation and range reduction.
- [FINUFFT] `src/spreadinterp.cpp::setup_spreader` (Apache-2.0) — the implemented width rule `W = ceil(-log10(eps/10))` at σ = 2.
- [jax-finufft] `lib/jax_finufft_cpu.cc` (Apache-2.0) — one `makeplan` per custom call, one `setpts` per point set.

**Evidence.** The measurements quoted above are from the 2026-09-02 review run (jax 0.9.2, jax-finufft 1.3.1,
ducc0 0.41.0 as oracle, macOS arm64 10-core, float64 unless stated). The review document is kept outside the
repository and is not to be committed; the definition-of-done tests below are the reproduction.

**Comment by chrisfinlay** [comment]
Scope addition: there is a **second** cancellation site, at call time rather than planning time, and this issue is its natural home.

The Context above covers `sqrt(1 - l^2 - m^2) - 1` cancelling near the **phase centre** (n -> 1, so n-1 -> 0). An independent review of the #11 branch found the mirror-image problem near the **horizon** (n -> 0): `wgridder.vis2dirty` reconstructs n as

```python
n_grid = (plan.n_minus_1 + 1.0).astype(dirty_per_chan.real.dtype)
```

before applying the `1/n` adjoint factor. For a pixel near the horizon `n_minus_1` is close to `-1`, so `n_minus_1 + 1` is exactly the cancelling subtraction this issue exists to remove — and the result is then used as a *divisor*, so the relative error is inherited directly by the output. Measured against ducc0 with a float32 plan at `epsilon = 1e-4`:

| pixel n | relative adjoint error |
|---|---|
| 1e-5 | ~1.3e-3 |
| 1e-6 | ~1.2e-2 |

Both are valid pixels inside the unit disc, and both are far outside the requested accuracy.

Suggested addition to the implementation plan:

4. Store the n grid itself as a plan leaf (`n_grid`), computed stably on the host in float64 alongside `n_minus_1` and cast at the end, and have `vis2dirty` divide by that leaf instead of reconstructing `n_minus_1 + 1.0` at call time. This is a new plan field, so it needs the full plan-field checklist treatment from AGENTS.md section 4: the dataclass, `_plan_aux` (as a leaf, not aux), `_plan_unflatten`, the pytree registration, and the leaf-count assertion in `test_plan_is_a_jax_pytree`.

5. Definition of done gains a near-horizon float32 adjoint regression at n = 1e-5 and n = 1e-6.

Note the interaction with #23 (plan storage), which reworks the plan leaves: whichever lands second inherits the other s field list.

**Comment by chrisfinlay** [comment]
## The float32 premise does not hold, but there is a real defect underneath it

Measured on `main` at `190a2e4`, `plan.n_minus_1` against the cancellation-free form evaluated in
`numpy.longdouble`, 64² grid, over the whole grid:

| pixsize (rad) | innermost \|n−1\| | dtype | max abs err | max rel err |
|---|---|---|---|---|
| 5e-3 | 1.250e-05 | float64 | 8.066e-17 | 4.437e-12 |
| 5e-3 | 1.250e-05 | float32 | 9.215e-10 | 1.068e-05 |
| 1e-4 | 5.000e-09 | float64 | 7.658e-17 | 1.363e-08 |
| 1e-4 | 5.000e-09 | float32 | 4.510e-13 | 1.059e-05 |
| 1e-6 | 5.000e-13 | float64 | 8.302e-17 | 8.890e-05 |
| 1e-8 | 5.000e-17 | float64 | 8.318e-17 | **1.220** |

**The issue says the value is "exactly 0 for the innermost pixels in float32". It is not, at any
pixel scale tested** — zero off-centre pixels are exactly zero at either dtype. The reason is that
`_n_minus_1_grid` (`planning.py:808`) computes the grid in **full float64 whatever the plan dtype**
and narrows once at the end, which its docstring states explicitly. The float32 cancellation the
issue describes cannot occur, because no part of this is evaluated in float32.

**But the cancellation is real in float64**, and at fine pixel scales it is severe: 8.9e-05 relative
at 1e-6 rad (≈0.21 arcsec) and 122% at 1e-8 rad (≈2 mas). Those are VLBI-realistic pixel sizes, so
this is not a purely theoretical regime. Note the last two rows are identical across dtypes — once
`n−1` falls below ~1e-13 the float64 computation dominates and the narrowing cast contributes
nothing.

## Why it has no observable consequence today

`n−1` enters through the w-phase `exp(2πi w (n−1))`, which depends on the **absolute** error, and
that is bounded at **~8.3e-17 across every row of the table** — it does not grow as the relative
error explodes, because the value itself is shrinking. The resulting phase error is
`2π · w · 8.3e-17`; reaching even 1e-6 rad of phase would need `w ≈ 2e9` wavelengths, far outside
any terrestrial baseline. The relative error is large only where `n−1` is negligible.

So the ranking in this issue is inverted: it is filed under accuracy with a float32 rationale, and
it is neither. It is a latent numerical wart that no current configuration can observe.

## What that means for the fix

The rewrite is still worth doing, for a different reason than the one stated, and it is unusually
safe: since the absolute error is ≤ 1.7e-16, the stable form changes `n−1` by less than a float64
ulp of the quantities it feeds, so **no measured number anywhere in the suite will move**. It is
cheap, monotone-improving, and cannot regress anything.

Suggested rewrite of the issue:

1. Drop the float32 rationale and the "exactly 0" claim.
2. State the real one: relative accuracy of `n−1` degrades without bound as `pixsize → 0`,
   reaching 122% at 1e-8 rad, and the standard rewrite `√(1−x) − 1 = −x/(√(1−x) + 1)`
   ([Higham2002] Ch. 1) fixes it at no cost.
3. Gate it on **relative** error against `numpy.longdouble` at a fine pixel scale — 1e-6 or 1e-8
   rad, where the current form fails and the stable one passes. At 5e-3 both pass, so a test at
   realistic imaging pixel sizes would not bite.
4. Drop item 2 of the implementation plan: `_utils.py` on `main` already contains only
   `SPEED_OF_LIGHT`. The three dead helpers were removed by an earlier PR, so half this issue is
   done.

## Definition of done, revised

- [ ] Relative error against `numpy.longdouble` below 1e-12 at pixsize 1e-8 rad (currently 1.22).
- [ ] Reverting to `sqrt(1 - x) - 1` fails that test (verify locally).
- [ ] No other measured number in the suite changes, confirming the change is below ulp downstream.
